picard/ui/searchdialog.py

```diff
@@ -0,0 +1,466 @@
+
+        QtGui.QDialog.accept(self)
+
+    def reject(self):
+        if getattr(self, "table", None):
+            self.save_state(True)
+        else:
+            self.save_state(False)
+
+        QtGui.QDialog.reject(self)
+
+    def restore_state(self):
+        size = config.persist["searchdialog_window_size"]
+        if size:
+            self.resize(size)
+        self.search_box.restore_checkbox_state()
+
+    def restore_table_header_state(self):
+        header = self.table.horizontalHeader()
+        state = config.persist["searchdialog_header_state"]
+        if state:
+            header.restoreState(state)
+        header.setResizeMode(QtGui.QHeaderView.Interactive)
+
+    def save_state(self, table_loaded=True):
+        """Saves dialog state i.e. window size, checkbox state, and table
+        header size.
+
+        Args:
+            table_loaded -- Whether table widget is loaded or not
+        """
+        if table_loaded:
+            self.save_table_header_state()
+        config.persist["searchdialog_window_size"] = self.size()
+        self.search_box.save_checkbox_state()
+
+    def save_table_header_state(self):
+        state = self.table.horizontalHeader().saveState()
+        config.persist["searchdialog_header_state"] = state
+
+
+class TrackSearchDialog(SearchDialog):
+
+    def __init__(self, parent):
+        super(TrackSearchDialog, self).__init__(parent)
+        self.file_ = None
+        self.setWindowTitle(_("Track Search Results"))
+        self.table_headers = [
+                _("Name"),
+                _("Length"),
+                _("Artist"),
+                _("Release"),
+                _("Date"),
+                _("Country"),
+                _("Type")
+        ]
+
+    def search(self, text):
+        """Performs search using query provided by the user."""
+        self.retry_params = (self.search, text)
+        self.search_box.search_edit.setText(text)
+        self.show_progress()
+        self.tagger.xmlws.find_tracks(self.handle_reply,
+                query=text,
+                search=True,
+                limit=25)
+
+    def load_similar_tracks(self, file_):
+        """Performs search by using existing metadata information
+        from the file."""
+        self.retry_params = (self.load_similar_tracks, file_)
+        self.file_ = file_
+        metadata = file_.orig_metadata
+        query = {
+                'track': metadata['title'],
+                'artist': metadata['artist'],
+                'release': metadata['album'],
+                'tnum': metadata['tracknumber'],
+                'tracks': metadata['totaltracks'],
+                'qdur': str(metadata.length / 2000),
+                'isrc': metadata['isrc'],
+        }
+        if config.setting["use_adv_search_syntax"]:
+            # Display the query in advance syntax format.
+            query_str = ' '.join(['%s:(%s)' % (item, value) for item, value in query.iteritems() if value])
```

Comment:
`value` needs to be passed through `_escape_lucene_query` from `picard.webservice`  here. Otherwise, the query string that's set in the search box could contain unescaped special characters.